himbaechel/uarch/gatemate/pll.cc

```diff
@@ -0,0 +1,352 @@
+            tmp.f_core = f_core_local;
+            tmp.M1 = M1;
+            tmp.M2 = M2;
+            tmp.weight = abs(f_core_local - f_core) * OUT_FREQ_OPT_FACTOR;
+            res_arr.push_back(tmp);
+
+            // frequency match
+            if (abs(f_core_local - f_core) <= f_core_delta) {
+                auto it = std::min_element(
+                        res_arr.begin(), res_arr.end(),
+                        [](const PllCfgRecord &a, const PllCfgRecord &b) { return a.weight < b.weight; });
+                size_t index = std::distance(res_arr.begin(), it);
+                setting.M1 = res_arr[index].M1;
+                setting.M2 = res_arr[index].M2;
+                setting.f_core = res_arr[index].f_core;
+                setting.f_core_delta = abs(res_arr[index].f_core - f_core);
+                setting.core_weight = res_arr[index].weight + setting.M1 + setting.M2 + setting.N1 + setting.N2 +
+                                      setting.K + setting.PDIV1;
+                return;
+            }
+        }
+    }
+}
+
+void get_DCO_ext_feedback(double f_core, double f_ref, PllCfgRecord &setting, int f_dco_min, int f_dco_max,
+                          double f_dco, double max_input_freq)
+{
+    std::vector<PllCfgRecord> res_arr;
+
+    for (int M1 = 1; M1 <= 64; M1++) {
+        for (int M2 = 1; M2 <= 1024; M2++) {
+            if (((M1 * M2) % 2) != 0) {
+                // M1*M2-->even number(except one);
+                // this is important to ensure 90 degree phase shifting of clock output
+                if (M1 * M2 != 1)
+                    continue;
+            }
+            double f_dco_local = (f_ref / setting.K) * 2 * setting.PDIV1 * setting.N1 * setting.N2 * M1 * M2;
+            if (f_dco_local > f_dco_max)
+                break; // upper limit
+            if ((f_dco_local < f_dco_min) || (f_dco_local > f_dco_max))
+                continue; // dco out of range limit
+            if (f_dco_local / setting.PDIV1 > max_input_freq)
+                continue; // M1 input freq limit
+            if ((f_dco_local / setting.PDIV1) / M1 > max_input_freq / 2)
+                continue; // M2 input freq limit
+
+            PllCfgRecord tmp;
+            tmp.f_dco = f_dco_local;
+            tmp.M1 = M1;
+            tmp.M2 = M2;
+            tmp.weight = f_dco_local + getDCO_optimized_value(f_dco_min, f_dco_max, f_dco_local);
+            res_arr.push_back(tmp);
+        }
+    }
+
+    auto it = std::min_element(res_arr.begin(), res_arr.end(),
+                               [](const PllCfgRecord &a, const PllCfgRecord &b) { return a.weight < b.weight; });
+    size_t index = std::distance(res_arr.begin(), it);
+    setting.M1 = res_arr[index].M1;
+    setting.M2 = res_arr[index].M2;
+    setting.f_dco = res_arr[index].f_dco;
+}
+}; // namespace
+
+NEXTPNR_NAMESPACE_BEGIN
+
+PllCfgRecord GateMatePacker::get_pll_settings(double f_ref, double f_core, int mode, int low_jitter, bool pdiv0_mux,
```

Comment:
not needed for this PR, but given all the calculations being done here, it would be nice to do clock constraint propagation too

Reply:
Addressed this one as well